x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/preprocessing/Entropy.java

```diff
@@ -0,0 +1,219 @@
+            NAME.getPreferredName(),
+            lenient,
+            (a, c) -> new Entropy((String)a[0],
+                (String)a[1],
+                a[2] == null ? c.isCustomByDefault() : (Boolean)a[2]));
+        parser.declareString(ConstructingObjectParser.constructorArg(), FIELD);
+        parser.declareString(ConstructingObjectParser.constructorArg(), FEATURE_NAME);
+        parser.declareBoolean(ConstructingObjectParser.optionalConstructorArg(), CUSTOM);
+        return parser;
+    }
+
+    public static Entropy fromXContentStrict(XContentParser parser, PreProcessorParseContext context) {
+        return STRICT_PARSER.apply(parser, context == null ?  PreProcessorParseContext.DEFAULT : context);
+    }
+
+    public static Entropy fromXContentLenient(XContentParser parser, PreProcessorParseContext context) {
+        return LENIENT_PARSER.apply(parser, context == null ?  PreProcessorParseContext.DEFAULT : context);
+    }
+
+    private final String field;
+    private final String featureName;
+    private final boolean custom;
+
+    public Entropy(String field, String featureName, boolean custom) {
+        this.field = ExceptionsHelper.requireNonNull(field, FIELD);
+        this.featureName = ExceptionsHelper.requireNonNull(featureName, FEATURE_NAME);
+        this.custom = custom;
+    }
+
+    public Entropy(StreamInput in) throws IOException {
+        this.field = in.readString();
+        this.featureName = in.readString();
+        this.custom = in.readBoolean();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(field);
+        out.writeString(featureName);
+        out.writeBoolean(custom);
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this);
+    }
+
+    @Override
+    public List<String> inputFields() {
+        return Collections.singletonList(field);
+    }
+
+    @Override
+    public List<String> outputFields() {
+        return Collections.singletonList(featureName);
+    }
+
+    @Override
+    public void process(Map<String, Object> fields) {
+        Object value = fields.get(field);
+        if (value == null) {
+            return;
+        }
+        // TODO should we support an array input for entropy of entries?
+        // TODO should entropy also optionally include length as a feature?
+        final String stringValue = value.toString();
+        double entropy = entropyCalc(stringValue);
+        fields.put(featureName, entropy);
+    }
+
+    double entropyCalc(String stringValue) {
+        final double len = stringValue.length();
+        // This provides a nice optimization for throughput.
+        // The entire basic latin code block fits is within 0x0000..0x007F.
+        // So, this provides a good optimization for those types of characters
+        // We still have a sparse hashmap for the rest.
+        int[] vals = new int[OPTIMIZATION_SIZE];
```

Comment:
One might suspect that this optimization is not worth it, but at the small expense of 127*4 bytes, we get almost double the speed.

```
# Benchmark on number of microseconds per op.
# This benchmark was the current hybrid approproach vs just a pre-allocated HashMap
# This is the average speed per ops for going through many strings of various sizes and languages
Benchmark                     Mode  Cnt    Score   Error  Units
EntropyBenchmark.hybrid       avgt   15  242.901 ± 4.729  us/op
EntropyBenchmark.unoptimized  avgt   15  405.232 ± 3.833  us/op
```